raft/src/main/java/org/apache/kafka/snapshot/RecordsSnapshotReader.java

```diff
@@ -121,9 +122,22 @@ private Optional<Batch<T>> nextBatch() {
             Batch<T> batch = iterator.next();
 
             if (!lastContainedLogTimestamp.isPresent()) {
-                // The Batch type doesn't support returning control batches. For now lets just use
-                // the append time of the first batch
-                lastContainedLogTimestamp = OptionalLong.of(batch.appendTimestamp());
+                // This must be the first batch which is expected to be a control batch with one record for
+                // the snapshot header.
+                if (batch.controlRecords().isEmpty()) {
+                    throw new IllegalStateException("First batch is not a control batch with at least one record");
```

Comment:
This will break compatibility with older snapshot versions. Certainly it breaks compatibility with the oldest raft snapshots, which had no control records. I don't recall the exact version where we added the initial control record. You need to find that version and then we can discuss whether breaking compatibility here is acceptable. If it is, we need to have an appropriate error message that talks about why the older version is not supported and which versions are supported.

Reply:
It does break backward compatibility for Kafka KRaft versions less than 3.3.0. Support for header and footer in snapshot was added in Kafka 3.3.0: https://github.com/apache/kafka/pull/10899.

We agree to break backwards compatibility for KRaft cluster in version 3.3.0. It is not possible for a KRaft cluster less than 3.3.0 to upgrade to a version later than 3.3.0. So it is safe to assume that every cluster running this version has a footer and header in the snapshot file.

Snapshot header and footer are not the only example of where we broke backwards compatibility in KRaft 3.3.0.

Reply:
Are you thinking of the 2.8 -> 3.0 compatibility break? I'm not aware of any such break between 3.2 and 3.3

Reply:
Yes. I got confused with the version numbers. Snapshot header and footer was added in 3.0.0:
```bash
$ git show --summary  d3ec9f940cc4402270cba3ad159ec64a9676a385
commit d3ec9f940cc4402270cba3ad159ec64a9676a385
Author: Niket <niket-goel@users.noreply.github.com>
Date:   Tue Jun 29 09:37:20 2021 -0700

    KAFKA-12952 Add header and footer records for raft snapshots (#10899)

    Add header and footer records for raft snapshots. This helps identify when the snapshot
    starts and ends. The header also contains a time.  The time field is currently set to 0.
    KAFKA-12997 will add in the necessary wiring to use the correct timestamp.

    Reviewers: Jose Sancio <jsancio@gmail.com>, Colin P. McCabe <cmccabe@apache.org>

 create mode 100644 clients/src/main/resources/common/message/SnapshotFooterRecord.json
 create mode 100644 clients/src/main/resources/common/message/SnapshotHeaderRecord.json
```
```bash
jsancio@jerry:~/work/kafka$ git tag --contains d3ec9f940cc4402270cba3ad159ec64a9676a385
3.0.0
3.0.0-rc0
3.0.0-rc1
3.0.0-rc2
3.0.1
3.0.1-rc0
3.0.2
3.0.2-rc0
3.1.0
3.1.0-rc0
3.1.0-rc1
3.1.1
3.1.1-rc0
3.1.1-rc1
3.1.2
3.1.2-rc0
3.2.0
3.2.0-rc0
3.2.0-rc1
3.2.1
3.2.1-rc2
3.2.1-rc3
3.2.2
3.2.2-rc0
3.2.3
3.2.3-rc0
3.3.0
3.3.0-rc1
3.3.0-rc2
3.3.1
3.3.1-rc0
3.3.2
3.3.2-rc1
3.4.0
3.4.0-rc2
```